# MaRS Laboratory SOP

The Malaria Resistance Surveillance or MaRS pipline, consists of a standardized laboratory and bioinformatics workflow for identifying variants in *Plasmodium falciparum* genomes conferring drug resistance. By streamlinging the data QC, alignment, and variant calling process, MaRS hopes to reduce time of analysis, as well allow for flexibilty of protocol used for the BGS analysis.

Version 3.3 (modified November 2, 2017)

*If you end up using MaRS in your workflow, please cite this [study](https://www.ncbi.nlm.nih.gov/pubmed/29439965):*

```
Next-Generation Sequencing and Bioinformatics Protocol for Malaria Drug Resistance Marker Surveillance.

Talundzic E1, Ravishankar S2, Kelley J3, Patel D4, Plucinski M4, Schmedes S4,5, Ljolje D3, Clemons B6, 
Madison-Antenucci S6, Arguin PM4, Lucchi NW4, Vannberg F2, Udhayakumar V4.

Antimicrob Agents Chemother. 2018 Mar 27;62(4). pii: e02474-17. doi: 10.1128/AAC.02474-17. Print 2018 Apr.
```


## MaRS Pipeline 

#### MaRS consists of two standardized workflows:
  * [Laboratory SOP](https://github.com/CDCgov/MaRS/tree/master/lab_sop/MaRSsop.md)
    * Latest version: Version 3.3 (modified November 2, 2017)
      * Currently validated for use with *Plasmodium falciparum* infected whole-blood samples
    * Table of Contents
      * Introduction/General Overview
      * Materials and Equipment
      * NGS Library Preparation Workflow
      * PET PCR Sample QC
      * Gene PCR Enrichment
      * Electrophoresis
      * SequalPrep Normalization and Purification
      * Tagment Genomic DNA
      * Library Amplification and Index PCR
      * Library Clean up
      * Library Pooling, Quantification, and Normalization
      * Library Denaturing and MiSeq Sample Loading
      * Supporting Information
  * [Bioinformatic pipeline](https://github.com/CDCgov/MaRS)
    * See MaRS README for bioinformatics pipeline protocol

### Contact Information ###

* For questions or other inquiries, please contact Eldin Talundzic (etalundzic@cdc.gov).
